### PR TITLE
MISUV-7608-prepopulate-data-from-session

### PR DESCRIPTION
**Pull request by simondrugan16**
PR for ticket to pre-populate data on pages when we have it available in the session, but also the has journey completed flag is false ^^

### CODE REVIEW

**Comment by avelx** [comment]
@simondrugan16  - if you can raise PR out of main next time as there are quite a lot of old commits in this PR

File: app/controllers/claimToAdjustPoa/AmendablePOAController.scala [hunk 1]
```diff
@@ -49,26 +50,44 @@ class AmendablePOAController @Inject()(val authorisedFunctions: AuthorisedFuncti
...
+            Future.failed(ex)
+        }
         } recover {
           case ex: Exception =>
             Logger("application").error(s"Unexpected error: ${ex.getMessage} - ${ex.getCause}")
             showInternalServerError(isAgent)
         }
     }
+
+  private def handleSession(implicit hc: HeaderCarrier): Future[Either[Throwable, Unit]] = {
```
Comment (avelx) [inline app/controllers/claimToAdjustPoa/AmendablePOAController.scala:80]: :thinking: - this method is a good candidate to be moved in the separate trait for re-use in other page in ClaimToAdjust the flow;
:wrench: - please add logging, especially around error cases;

File: app/forms/mappings/Formatters.scala [hunk 2]
```diff
@@ -95,4 +95,30 @@ trait Formatters {
...
+
+      private val baseFormatter = stringFormatter(emptyErrorMessageKey)
+
+      override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], BigDecimal] =
+        baseFormatter
+          .bind(key, data)
+          .flatMap {
+            case s if s.isEmpty =>
+              Left(Seq(FormError(key, emptyErrorMessageKey)))
+            case s if !s.matches(validNumeric) || !s.matches(is2dp) =>
```
Comment (avelx) [inline app/forms/mappings/Formatters.scala:113]: it should be possible to merge this regExp expression into one, are we actually testing this logic ?

File: it/test/controllers/claimToAdjustPoa/AmendablePOAControllerISpec.scala [hunk 3]
```diff
@@ -53,6 +57,64 @@ class AmendablePOAControllerISpec extends ComponentSpecBase {
           httpStatus(OK)
         )
       }
+      "User is authorised and there is an active session and journeyCompleted = false" in {
```
Comment (avelx) [inline it/test/controllers/claimToAdjustPoa/AmendablePOAControllerISpec.scala:60]: when writing test, please make sure there is at least one scenario with amount containing the fractional part ie: £250.13 etc

File: it/test/controllers/claimToAdjustPoa/EnterPoAAmountControllerISpec.scala [hunk 4]
```diff
@@ -143,6 +143,34 @@ class EnterPoAAmountControllerISpec extends ComponentSpecBase {
...
+        And("I wiremock stub financial details for multiple years with POAs")
+        IncomeTaxViewChangeStub.stubGetFinancialDetailsByDateRange(testNino, s"${testTaxYear - 1}-04-06", s"$testTaxYear-04-05")(
+          OK, testValidFinancialDetailsModelJson(2000, 2000, (testTaxYear - 1).toString, testDate.toString, poaRelevantAmount = Some(2000))
+        )
+        IncomeTaxViewChangeStub.stubGetFinancialDetailsByDateRange(testNino, s"${testTaxYear - 2}-04-06", s"${testTaxYear - 1}-04-05")(
+          OK, testValidFinancialDetailsModelJson(2000, 2000, (testTaxYear - 1).toString, testDate.toString, poaRelevantAmount = Some(2000))
+        )
+
+        And("A session has been created")
+        sessionService.setMongoData(Some(PoAAmendmentData(None, Some(BigDecimal(3333.33)))))
```
Comment (avelx) [inline it/test/controllers/claimToAdjustPoa/EnterPoAAmountControllerISpec.scala:163]: :+1:

Comment (avelx) [inline it/test/controllers/claimToAdjustPoa/EnterPoAAmountControllerISpec.scala:163]: I think worth to add _await_ for each call to sessionService?

File: it/test/controllers/claimToAdjustPoa/SelectYourReasonControllerISpec.scala [hunk 5]
```diff
@@ -100,6 +100,34 @@ class SelectYourReasonControllerISpec extends ComponentSpecBase {
...
+        And("Financial details for multiple years with POAs")
+        IncomeTaxViewChangeStub.stubGetFinancialDetailsByDateRange(testNino, s"${testTaxYear - 1}-04-06", s"$testTaxYear-04-05")(
+          OK, testValidFinancialDetailsModelJson(2000, 2000, (testTaxYear - 1).toString, testDate.toString, poaRelevantAmount = Some(3000))
+        )
+        IncomeTaxViewChangeStub.stubGetFinancialDetailsByDateRange(testNino, s"${testTaxYear - 2}-04-06", s"${testTaxYear - 1}-04-05")(
+          OK, testValidFinancialDetailsModelJson(2000, 2000, (testTaxYear - 1).toString, testDate.toString, poaRelevantAmount = Some(3000))
+        )
+
+        And("A session has been created")
+        sessionService.setMongoData(Some(PoAAmendmentData(Some(MainIncomeLower))))
```
Comment (avelx) [inline it/test/controllers/claimToAdjustPoa/SelectYourReasonControllerISpec.scala:122]: please add await

File: app/forms/mappings/Formatters.scala (hunk 2 above)
Comment (simondrugan16) [reply app/forms/mappings/Formatters.scala:113]: This is something I just refactored away from the form that Daniel made. He took this from Submissions project do I'm not sure of the original intent of the ticket

**Review by avelx — APPROVED** [review]

**Pull request merged**